File: typescript/packages/core/src/workspace/executor/builtins/shared.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult } from '../../../io/types.ts'
import type { SessionView } from '../../../ops/types.ts'
import type { PolicyDenied } from '../../../policy/errors.ts'
import type { ArithError } from '../../../shell/errors.ts'
import { PathSpec, wordText } from '../../../types.ts'
import { mountKey } from '../../../utils/key_prefix.ts'
import { resolvePath } from '../../../utils/path.ts'
import { rstripSlash } from '../../../utils/slash.ts'
import type { Namespace } from '../../mount/namespace/namespace.ts'
import { ExecutionNode } from '../../types.ts'
import { IDENTIFIER_RE } from './constants.ts'
import type { Result } from './types.ts'

const ENC = new TextEncoder()

interface ResultInit {
  out?: Uint8Array | null
  exitCode?: number
  stderr?: string
  io?: IOResult
}

/**
 * Build the (stream, IOResult, ExecutionNode) triple builtins return.
 *
 * @param cmd - command name recorded on the ExecutionNode.
 * @param init - `out` stdout payload; `exitCode` for both results; `stderr`
 *   error text encoded onto both; `io` a prebuilt IOResult to reuse (e.g.
 *   carrying writes), whose exitCode/stderr are overwritten.
 */
export function result(cmd: string, init: ResultInit = {}): Result {
  const exitCode = init.exitCode ?? 0
  const err =
    init.stderr !== undefined && init.stderr !== '' ? ENC.encode(init.stderr) : new Uint8Array()
  const io = init.io ?? new IOResult()
  io.exitCode = exitCode
  if (err.length > 0) io.stderr = err
  return [init.out ?? null, io, new ExecutionNode({ command: cmd, exitCode, stderr: err })]
}

export function ok(cmd: string, out?: Uint8Array | null): Result {
  return result(cmd, { out: out ?? null })
}

export function fail(cmd: string, message: string, exitCode = 1): Result {
  return result(cmd, { exitCode, stderr: message })
}

/**
 * Close an operand loop: exit 1 with joined stderr when any operand failed,
 * exit 0 otherwise.
 *
 * @param cmd - command name.
 * @param errors - per-operand error messages collected so far.
 * @param io - prebuilt IOResult to reuse (e.g. carrying writes).
 */
export function finish(cmd: string, errors: string[], io?: IOResult): Result {
  const carried = io !== undefined ? { io } : {}
  if (errors.length > 0) {
    return result(cmd, { exitCode: 1, stderr: errors.join(''), ...carried })
  }
  return result(cmd, carried)
}

/**
 * A non-path operand's text (a mode or owner spec the classifier may have
 * wrapped as a path).
 *
 * @param arg - a classified command part.
 */
export function operandText(arg: string | PathSpec): string {
  return arg instanceof PathSpec ? arg.virtual : arg
}

/**
 * A path operand as an absolute virtual path.
 *
 * @param arg - a classified command part.
 * @param cwd - session working directory for relative operands.
 */
export function absPath(arg: string | PathSpec, cwd: string): string {
  if (arg instanceof PathSpec) return arg.virtual
  return resolvePath(arg, cwd)
}

function allKnown(chars: string, known: string): boolean {
  for (const c of chars) if (!known.includes(c)) return false
  return true
}

/**
 * Split leading single-letter flags, permissively.
 *
 * A token containing any unknown letter is kept as an operand instead of
 * erroring (`ln`/`readlink` behavior).
 *
 * @param args - args after the command name.
 * @param known - accepted single-letter flags.
 * @returns [flags, operands].
 */
export function splitFlags(
  args: (string | PathSpec)[],
  known: string,
): [Set<string>, (string | PathSpec)[]] {
  const flags = new Set<string>()
  const operands: (string | PathSpec)[] = []
  let parsing = true
  for (const arg of args) {
    const s = operandText(arg)
    if (parsing && s === '--') {
      parsing = false
      continue
    }
    if (parsing && s !== '-' && s.length >= 2 && s.startsWith('-') && allKnown(s.slice(1), known)) {
      for (const c of s.slice(1)) flags.add(c)
      continue
    }
    parsing = false
    operands.push(arg)
  }
  return [flags, operands]
}

export interface SplitValueFlags {
  flags: Set<string>
  values: Map<string, string>
  operands: (string | PathSpec)[]
  bad: string | null
}

/**
 * Split leading flags where some take a value (`-t STAMP`), strictly: an
 * unknown letter is reported instead of tolerated.
 *
 * @param args - args after the command name.
 * @param boolean - single-letter flags with no value.
 * @param valued - single-letter flags that consume the next arg.
 */
export function splitValueFlags(
  args: readonly (string | PathSpec)[],
  boolean: string,
  valued: string,
): SplitValueFlags {
  const flags = new Set<string>()
  const values = new Map<string, string>()
  const operands: (string | PathSpec)[] = []
  let parsing = true
  let i = 0
  while (i < args.length) {
    const arg = args[i]
    if (arg === undefined) break
    const s = operandText(arg)
    if (parsing && s === '--') {
      parsing = false
      i += 1
      continue
    }
    if (parsing && s !== '-' && s.length >= 2 && s.startsWith('-') && !s.startsWith('--')) {
      const body = s.slice(1)
      for (let j = 0; j < body.length; j++) {
        const c = body.charAt(j)
        if (boolean.includes(c)) {
          flags.add(c)
          continue
        }
        // A valued flag consumes the rest of the token (-tSTAMP) or the next
        // argument (-t STAMP); those trailing chars are its value, not flags,
        // so validation must stop here rather than pre-scanning the token.
        if (!valued.includes(c)) {
          return { flags, values, operands, bad: c }
        }
        const rest = body.slice(j + 1)
        if (rest.length > 0) {
          values.set(c, rest)
        } else if (i + 1 < args.length) {
          i += 1
          const nxt = args[i]
          if (nxt !== undefined) values.set(c, wordText(nxt))
        }
        break
      }
      i += 1
      continue
    }
    parsing = false
    operands.push(arg)
    i += 1
  }
  return { flags, values, operands, bad: null }
}

/**
 * Coerce operands to PathSpec and expand glob patterns per mount.
 *
 * A pattern spec only exists for a mounted word (classification gates it), so
 * the lookup propagates on a miss; a backend with no glob keeps the literal
 * spec.
 *
 * @param namespace - addressing authority (mount lookup).
 * @param operands - positional operands.
 */
export async function expandOperands(
  namespace: Namespace,
  operands: readonly (string | PathSpec)[],
): Promise<PathSpec[]> {
  const out: PathSpec[] = []
  for (const item of operands) {
    const spec = item instanceof PathSpec ? item : PathSpec.fromStrPath(item)
    if (spec.pattern !== null) {
      const mount = namespace.mountFor(spec.virtual)
      if (mount.vfs.glob !== undefined) {
        const prefix = rstripSlash(mount.prefix)
        const withPrefix = new PathSpec({
          virtual: spec.virtual,
          directory: spec.directory,
          pattern: spec.pattern,
          resolved: spec.resolved,
          vfsPath: mountKey(spec.virtual, prefix),
        })
        const expanded = await mount.expandGlob([withPrefix], prefix)
        for (const p of expanded) if (p instanceof PathSpec) out.push(p)
        continue
      }
    }
    out.push(spec)
  }
  return out
}

/**
 * The gated session view this builtin writes through.
 *
 * Every session write goes through the workspace's gated view, which is
 * what makes `preSession` rules enforceable; this used to fall back to
 * an ungated view over the same session, so a caller that forgot to
 * thread one silently wrote past every policy. A write reached without
 * a view is a wiring bug, not a mode, so it throws.
 */
export function requireView(state: SessionView | null): SessionView {
  if (state === null) {
    throw new Error(
      "builtin reached a session write without the workspace's gated " +
        'session view; thread state from the executor arm',
    )
  }
  return state
}

/** Render a policy denial in the builtin's own voice. */
export function refusal(cmd: string, err: PolicyDenied): Result {
  const encoded = new TextEncoder().encode(`${err.message}\n`)
  return [
    null,
    new IOResult({ exitCode: 1, stderr: encoded }),
    new ExecutionNode({ command: cmd, exitCode: 1, stderr: encoded }),
  ]
}

/** Render the shell's own readonly refusal, checked before the door. */
export function readonlyRefusal(cmd: string, name: string): Result {
  const encoded = new TextEncoder().encode(`bash: ${name}: readonly variable\n`)
  return [
    null,
    new IOResult({ exitCode: 1, stderr: encoded }),
    new ExecutionNode({ command: cmd, exitCode: 1, stderr: encoded }),
  ]
}

/**
 * Render the `-i` coercion's arithmetic error as bash does.
 *
 * GNU voices it as the evaluator's own line, prefixed by the builtin and
 * the offending text (`bash: read: 1+: syntax error: operand expected`),
 * and fails the builtin with 1 while the variable keeps its old value,
 * which is what the door's copy-then-store already guarantees. A plain
 * assignment (`n=1+`) is fatal instead and is voiced by the executor
 * without a builtin name.
 */
export function arithRefusal(cmd: string, err: ArithError): Result {
  const encoded = new TextEncoder().encode(`bash: ${cmd}: ${err.message}\n`)
  return [
    null,
    new IOResult({ exitCode: 1, stderr: encoded }),
    new ExecutionNode({ command: cmd, exitCode: 1, stderr: encoded }),
  ]
}

/** Whether the word is a shell identifier. */
export function isValidName(name: string): boolean {
  return IDENTIFIER_RE.test(name)
}

/**
 * Whether the word is an optionally signed run of digits, which is what
 * `shift`, `return` and `exit` accept as their argument.
 */
export function isCountWord(word: string): boolean {
  const body = word.startsWith('-') || word.startsWith('+') ? word.slice(1) : word
  return /^\d+$/.test(body)
}

File: python/mirage/workspace/executor/builtins/shared.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.io import IOResult
from mirage.ops.types import SessionView
from mirage.policy import PolicyDenied
from mirage.shell.errors import ArithError
from mirage.types import PathSpec, word_text
from mirage.utils.path import resolve_path
from mirage.workspace.executor.builtins.constants import IDENTIFIER_RE
from mirage.workspace.executor.builtins.types import Result
from mirage.workspace.mount.namespace import Namespace
from mirage.workspace.types import ExecutionNode


def result(
    cmd: str,
    out: bytes | None = None,
    exit_code: int = 0,
    stderr: str | None = None,
    io: IOResult | None = None,
) -> Result:
    """Build the (stream, IOResult, ExecutionNode) triple builtins return.

    Args:
        cmd (str): command name recorded on the ExecutionNode.
        out (bytes | None): stdout payload, if any.
        exit_code (int): exit code for both IOResult and ExecutionNode.
        stderr (str | None): error text; encoded onto both results.
        io (IOResult | None): prebuilt IOResult to reuse (e.g. carrying
            writes); its exit_code/stderr are overwritten.
    """
    err = stderr.encode() if stderr else b""
    io = io if io is not None else IOResult()
    io.exit_code = exit_code
    if err:
        io.stderr = err
    return out, io, ExecutionNode(command=cmd, exit_code=exit_code, stderr=err)


def ok(cmd: str, out: bytes | None = None) -> Result:
    return result(cmd, out=out)


def fail(cmd: str, message: str, exit_code: int = 1) -> Result:
    return result(cmd, exit_code=exit_code, stderr=message)


def finish(cmd: str, errors: list[str], io: IOResult | None = None) -> Result:
    """Close an operand loop: exit 1 with joined stderr when any operand
    failed, exit 0 otherwise.

    Args:
        cmd (str): command name.
        errors (list[str]): per-operand error messages collected so far.
        io (IOResult | None): prebuilt IOResult to reuse (e.g. carrying
            writes).
    """
    if errors:
        return result(cmd, exit_code=1, stderr="".join(errors), io=io)
    return result(cmd, io=io)


def operand_text(arg: str | PathSpec) -> str:
    """A non-path operand's text (a mode or owner spec the classifier may
    have wrapped as a path).

    Args:
        arg (str | PathSpec): a classified command part.
    """
    return arg.virtual if isinstance(arg, PathSpec) else str(arg)


def abs_path(arg: str | PathSpec, cwd: str) -> str:
    """A path operand as an absolute virtual path.

    Args:
        arg (str | PathSpec): a classified command part.
        cwd (str): session working directory for relative operands.
    """
    if isinstance(arg, PathSpec):
        return arg.virtual
    return resolve_path(arg, cwd)


def split_flags(
    args: list[str | PathSpec],
    known: str,
) -> tuple[set[str], list[str | PathSpec]]:
    """Split leading single-letter flags, permissively.

    A token containing any unknown letter is kept as an operand instead
    of erroring (``ln``/``readlink`` behavior).

    Args:
        args (list[str | PathSpec]): args after the command name.
        known (str): accepted single-letter flags.

    Returns:
        tuple: (flags, operands).
    """
    flags: set[str] = set()
    operands: list[str | PathSpec] = []
    parsing = True
    for arg in args:
        s = operand_text(arg)
        if parsing and s == "--":
            parsing = False
            continue
        if (parsing and s != "-" and len(s) >= 2 and s.startswith("-")
                and all(c in known for c in s[1:])):
            flags.update(s[1:])
            continue
        parsing = False
        operands.append(arg)
    return flags, operands


def split_value_flags(
    args: list[str | PathSpec],
    boolean: str,
    valued: str,
) -> tuple[set[str], dict[str, str], list[str | PathSpec], str | None]:
    """Split leading flags where some take a value (``-t STAMP``),
    strictly: an unknown letter is reported instead of tolerated.

    Args:
        args (list[str | PathSpec]): args after the command name.
        boolean (str): single-letter flags with no value.
        valued (str): single-letter flags that consume the next arg.

    Returns:
        tuple: (bool flags, valued flags, operands, bad option or None).
    """
    flags: set[str] = set()
    values: dict[str, str] = {}
    operands: list[str | PathSpec] = []
    parsing = True
    i = 0
    while i < len(args):
        arg = args[i]
        s = operand_text(arg)
        if parsing and s == "--":
            parsing = False
            i += 1
            continue
        if parsing and s != "-" and len(s) >= 2 and s.startswith(
                "-") and not s.startswith("--"):
            body = s[1:]
            for j, c in enumerate(body):
                if c in boolean:
                    flags.add(c)
                    continue
                if c not in valued:
                    return flags, values, operands, c
                # A valued flag consumes the rest of the token (-tSTAMP)
                # or the next argument (-t STAMP).
                rest = body[j + 1:]
                if rest:
                    values[c] = rest
                elif i + 1 < len(args):
                    i += 1
                    values[c] = word_text(args[i])
                break
            i += 1
            continue
        parsing = False
        operands.append(arg)
        i += 1
    return flags, values, operands, None


async def expand_operands(
    namespace: Namespace,
    operands: list[str | PathSpec],
) -> list[PathSpec]:
    """Coerce operands to PathSpec and expand glob patterns per mount.

    Args:
        namespace (Namespace): addressing authority (mount lookup).
        operands (list[str | PathSpec]): positional operands.
    """
    out: list[PathSpec] = []
    for item in operands:
        spec = item if isinstance(item, PathSpec) else PathSpec.from_str_path(
            str(item))
        if spec.pattern:
            mount = namespace.mount_for(spec.virtual)
            expanded = await mount.expand_glob([spec],
                                               mount.prefix.rstrip("/"))
            out.extend(p for p in expanded if isinstance(p, PathSpec))
            continue
        out.append(spec)
    return out


def require_view(state: SessionView | None) -> SessionView:
    """The gated session view this builtin writes through.

    Every session write goes through the workspace's gated view, which
    is what makes ``pre_session`` rules enforceable; this used to fall
    back to an ungated view over the same session, so a caller that
    forgot to thread one silently wrote past every policy. A write
    reached without a view is a wiring bug, not a mode, so it raises.

    Args:
        state (SessionView | None): the caller's view, if threaded.

    Raises:
        RuntimeError: no view was threaded.
    """
    if state is None:
        raise RuntimeError(
            "builtin reached a session write without the workspace's gated "
            "session view; thread state= from the executor arm")
    return state


def refusal(cmd: str, exc: PolicyDenied) -> Result:
    """Render a policy denial in the builtin's own voice.

    Args:
        cmd (str): builtin name for the node.
        exc (PolicyDenied): the gate's refusal.
    """
    err = f"{exc.strerror}\n".encode()
    return None, IOResult(exit_code=1, stderr=err), ExecutionNode(command=cmd,
                                                                  exit_code=1,
                                                                  stderr=err)


def readonly_refusal(cmd: str, name: str) -> Result:
    """Render the shell's own readonly refusal, checked before the door.

    Args:
        cmd (str): builtin name for the node.
        name (str): the frozen variable.
    """
    err = f"bash: {name}: readonly variable\n".encode()
    return None, IOResult(exit_code=1, stderr=err), ExecutionNode(command=cmd,
                                                                  exit_code=1,
                                                                  stderr=err)


def arith_refusal(cmd: str, exc: ArithError) -> Result:
    """Render the ``-i`` coercion's arithmetic error as bash does.

    GNU voices it as the evaluator's own line, prefixed by the builtin
    and the offending text (``bash: read: 1+: syntax error: operand
    expected``), and fails the builtin with 1 while the variable keeps
    its old value, which is what the door's copy-then-store already
    guarantees. A plain assignment (``n=1+``) is fatal instead and is
    voiced by the executor without a builtin name.

    Args:
        cmd (str): builtin name for the node.
        exc (ArithError): the evaluator's refusal, text already led.
    """
    err = f"bash: {cmd}: {exc}\n".encode()
    return None, IOResult(exit_code=1, stderr=err), ExecutionNode(command=cmd,
                                                                  exit_code=1,
                                                                  stderr=err)


def is_valid_name(name: str) -> bool:
    """Whether the word is a shell identifier.

    Args:
        name (str): the word to test.
    """
    return IDENTIFIER_RE.fullmatch(name) is not None


def is_count_word(word: str) -> bool:
    """Whether the word is an optionally signed run of digits, which is
    what ``shift``, ``return`` and ``exit`` accept as their argument.

    Args:
        word (str): the word to test.
    """
    body = word[1:] if word[:1] in ("-", "+") else word
    return body.isdigit()
